### DISCUSSION TITLE
Bun all-in-one, with AssemblyScript as a webassembly language!

### DISCUSSION THREAD

**Discussion by junior-boop** [oven-sh/bun#6205] [Ideas]
Today, Bun.js is truly a game changer.

In past years, we were forced to choose the best tool from a vast number of options for our projects. Bun introduces native tools that are not only significantly faster than older alternatives but also easier to use. Some of them even require no configuration at all.

Today, Bun.js can run `TypeScript`, `JSX`, and` TSX` without any configuration. That’s why I propose to the Bun team to add support for the .as extension for AssemblyScript files, making Bun a compiler, a transpiler, and a WebAssembly environment.

### Why this idea?
WebAssembly is widely used today—on servers, in front-end applications for complex operations and design, and even to enhance the performance of graphical applications.

For JavaScript developers who want to experiment with WebAssembly without using Rust or C++, AssemblyScript is a great alternative due to its TypeScript-friendly syntax.

So my idea is to integrate a complete WebAssembly environment into Bun.js, with AssemblyScript as the entry point. To differentiate AssemblyScript files from JavaScript and TypeScript files, I suggest using the .as extension for all AssemblyScript files.

This is a dream of mine, and I would be thrilled if you respond favorably to my idea.

**Comment by MohsinDev369** [comment]
I would love to have this in bun.

**Comment by dsir-minute** [comment]
yes, go for it , vas y mon gars !

**Comment by m-i-z-u-k-i** [comment]
If they conside this they should also accept normal wasm files because assemblyscript has nothing to do with eebassembly on its own

**Comment by Miciurash** [comment]
+1 - this would be great.

**Comment by deepak0g** [comment]
Should start with wasm tbh

**Reply by MulverineX** [reply]
WASM **is** supported. If you mean compiling Bun/JavascriptCore itself to WASM or rolling QuickJS, IMO that's a bad idea, they should focus more on sandboxing/permissions at that point.

**Comment by cristiano-belloni** [comment]
AssemblyScript seems essentially dead. A good project to look at is porffor.dev, but it's in its infancy.

**Reply by MulverineX** [reply]
Define dead? ~~It supports WASM GC~~, there was a functional commit last month, and it has been sponsored by Shopify.

**Reply by MulverineX** [reply]
Its a good thing that they're not doing massive changes to the language anymore, that means its a stable and mature language to build things in.

**Reply by cristiano-belloni** [reply]
> Define dead? It supports WASM GC, there was a functional commit last month, and it has been sponsored by Shopify.

[This issue](https://github.com/AssemblyScript/assemblyscript/issues/2715) doesn't give me much safety on the project's future.
Commits are not terribly frequent (1 per month?) and WasmGC is not supported yet (from 2023). I really hope the project is not dead, but it doesn't give me incredibly safe vibes.

**Reply by MulverineX** [reply]
https://github.com/AssemblyScript/assemblyscript/issues/2808 guess I assumed wrong, apologies. Looks like the corporate sponsors of the project don't give a shit about pushing it forward and the FOSS maintainers basically are just providing life support and merging corporate optimization PRs. This is very unfortunate, I hope something changes.

**Comment by ShalokShalom** [comment]
https://spritely.institute/hoot/

Scheme, implemented in native WebAssembly garbage-collection, very lean and IMHO nicely designed.

**Reply by MulverineX** [reply]
A Lisp-alike language is not very attractive to most Typescript developers.

**Comment by ShalokShalom** [comment]
> So my idea is to integrate a complete WebAssembly environment into Bun.js

So, a friend of mine just did this. https://github.com/gaurav-Gosain/quickjs#readme

Note, this is very fresh; he could still spin this up very quickly. 
QuickJS is easier to implement, sure. 

Still, wazero enables integration of any program that can compile to wasm, with a few exceptions. 
TinyGo crashes with the syscalls; Go is implemented just fine.

**Reply by MulverineX** [reply]
Not sure whether this is relevant to the discussion

**Reply by ShalokShalom** [reply]
It shows a project that did exactly what this discussion is trying to do, just with a different JS implementation.

It probably took less time to implement that than this whole discussion around a dead language is going on. 🙄

**Comment by ShalokShalom** [comment]
You can turn JavaScript into WebAssembly with [porffor](https://github.com/CanadaHonk/porffor).

And for proper WebAssembly languages, there is [Hoot](https://spritely.institute/hoot/), which might serve as a proper extension language. 

It and [Moonbit](https://www.moonbitlang.com/) are currently the only two languages that use the native WebAssembly garbage collection, 
and don't need to bring their own runtime to reduce the deployment size.

**Reply by MulverineX** [reply]
`porffor` is a hobbist proof of concept, not something Bun needs to be rolling as an official integration imo. Hoot you already mentioned. Moonbit actually looks good, but again, not really relevant for Bun, some devs might want to use it anyway though.
